plugin/completion.py

```diff
@@ -72,274 +133,79 @@ def initialize(self) -> None:
-    def apply_additional_edits(self, additional_edits: List[Dict]) -> None:
-        edits = list(parse_text_edit(additional_edit) for additional_edit in additional_edits)
-        debug('applying additional edits:', edits)
-        self.view.run_command("lsp_apply_document_edit", {'changes': edits})
-        sublime.status_message('Applied additional edits for completion')
-
-    def handle_response(self, response: Optional[Union[Dict, List]]) -> None:
-        if self.state == CompletionState.REQUESTING:
-
-            completion_start = self.last_location
-            if position_is_word(self.view, self.last_location):
-                # if completion is requested in the middle of a word, where does it start?
-                word = self.view.word(self.last_location)
-                completion_start = word.begin()
-
-            current_word_start = self.view.sel()[0].begin()
-            if position_is_word(self.view, current_word_start):
-                current_word_region = self.view.word(current_word_start)
-                current_word_start = current_word_region.begin()
-
-            if current_word_start != completion_start:
-                debug('completion results for', completion_start, 'now at', current_word_start, 'discarding')
-                self.state = CompletionState.IDLE
-                return
-
-            _last_row, last_col = self.view.rowcol(completion_start)
-
-            response_items, response_incomplete = parse_completion_response(response)
-            self.response_items = response_items
-            self.response_incomplete = response_incomplete
-            self.completions = list(format_completion(item, last_col, settings) for item in self.response_items)
-
-            # if insert_best_completion was just ran, undo it before presenting new completions.
-            prev_char = self.view.substr(self.view.sel()[0].begin() - 1)
-            if prev_char.isspace():
-                if last_text_command == "insert_best_completion":
-                    self.view.run_command("undo")
+        flags = 0
+        if settings.only_show_lsp_completions:
+            flags |= sublime.INHIBIT_WORD_COMPLETIONS
+            flags |= sublime.INHIBIT_EXPLICIT_COMPLETIONS
 
-            self.state = CompletionState.APPLYING
-            self.view.run_command("hide_auto_complete")
-            self.run_auto_complete()
-        elif self.state == CompletionState.CANCELLING:
-            self.state = CompletionState.IDLE
-            if self.next_request:
-                prefix, locations = self.next_request
-                self.do_request(prefix, locations)
-        else:
-            debug('Got unexpected response while in state {}'.format(self.state))
+        if response_incomplete:
+            flags |= sublime.DYNAMIC_COMPLETIONS
+        completion_list.set_completions(items, flags)
 
     def handle_error(self, error: dict) -> None:
         sublime.status_message('Completion error: ' + str(error.get('message')))
```

Comment:
You should resolve the promise object during an error too. Return zero completions in this case.

Reply:
Done